File: backend/app/schemas/security.py
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.portal import COMPLETENESS_NOTICE
from app.schemas.records import SourceView

VerificationReason = Literal[
    "HASH_MISMATCH", "SEQUENCE_GAP", "LINK_MISMATCH", "CHECKPOINT_MISMATCH", "TRUNCATED"
]


class StrictSecurityModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class AuditEventView(StrictSecurityModel):
    schema_version: Literal[1]
    event_id: UUID
    stream_id: UUID
    sequence: int = Field(ge=1)
    event_type: str = Field(min_length=1, max_length=80)
    recorded_at: str
    occurred_at: str
    actor_id: UUID | None
    role_snapshot: str | None
    organization_id: UUID | None
    patient_ref: UUID | None
    source_org: UUID | None
    recipient_org: UUID | None
    resource_domain: str | None
    action: str = Field(min_length=1, max_length=80)
    decision: Literal["ALLOW", "DENY", "NOT_APPLICABLE"]
    reason_code: str = Field(min_length=1, max_length=80)
    policy_version: int | None
    consent_or_emergency_ref: UUID | None
    correlation_id: UUID
    outcome: Literal["SUCCEEDED", "DENIED", "ABORTED", "FAILED", "UNKNOWN"]
    context: dict[str, Any]
    justification_id: UUID | None
    justification_digest: str | None
    previous_hash: str = Field(pattern=r"^[0-9a-f]{64}$")
    event_hash: str = Field(pattern=r"^[0-9a-f]{64}$")


class EventCollection(StrictSecurityModel):
    items: list[AuditEventView] = Field(max_length=100)
    next_cursor: str | None
    correlation_id: UUID
    source: SourceView | None = None
    retrieved_at: str
    completeness_notice: str = Field(default=COMPLETENESS_NOTICE, min_length=1, max_length=400)


class VerifyChainRequest(StrictSecurityModel):
    trusted_checkpoint_id: UUID | None = None


class CheckpointView(StrictSecurityModel):
    stream_id: UUID
    sequence: int = Field(ge=0)
    head_hash: str = Field(pattern=r"^[0-9a-f]{64}$")
    created_at: str


class ChainVerification(StrictSecurityModel):
    stream_id: UUID
    status: Literal["VALID", "INVALID"]
    checked_from: int = Field(ge=0)
    checked_to: int = Field(ge=0)
    first_failing_sequence: int | None
    reason: VerificationReason | None
    checkpoint_comparison: Literal["MATCH", "MISMATCH", "NOT_PROVIDED"]
    checkpoint: CheckpointView
    verified_at: str
    limitations: list[str] = Field(max_length=10)
    correlation_id: UUID


class SecurityAlertView(StrictSecurityModel):
    id: UUID
    event_id: UUID
    stream_id: UUID
    rule_id: Literal["AR01", "AR02", "AR03", "AR04", "AR05", "AR06", "AR07", "AR08", "AR09"]
    severity: Literal["HIGH", "CRITICAL"]
    status: Literal[
        "REVIEW_REQUIRED", "IN_REVIEW", "RESOLVED_LEGITIMATE", "RESOLVED_SUSPECTED_MISUSE"
    ]
    actor_id: UUID
    organization_id: UUID | None
    patient_ref: UUID | None
    reason_code: str = Field(min_length=1, max_length=80)
    created_at: str
    reviewer_id: UUID | None
    resolution: str | None = Field(default=None, min_length=20, max_length=1000)
    version: int = Field(ge=1)


class AlertCollection(StrictSecurityModel):
    items: list[SecurityAlertView] = Field(max_length=100)
    next_cursor: str | None
    correlation_id: UUID
    source: SourceView | None = None
    retrieved_at: str
    completeness_notice: str = Field(default=COMPLETENESS_NOTICE, min_length=1, max_length=400)


class ReviewAlert(StrictSecurityModel):
    target_status: Literal["IN_REVIEW", "RESOLVED_LEGITIMATE", "RESOLVED_SUSPECTED_MISUSE"]
    explanation: str = Field(min_length=20, max_length=1000)
    expected_version: int = Field(ge=1)


class AlertResponse(StrictSecurityModel):
    alert: SecurityAlertView
    correlation_id: UUID

File: frontend/lib/api/contracts/security.ts
import { z } from "zod";
import { roleSchema, sourceSchema, type Role, type Source } from "./auth";
import { domainSchema } from "./records";

const uuid = z.uuid();
const utcDateTime = z.iso.datetime({ offset: false });
const hash = z.string().regex(/^[0-9a-f]{64}$/);

export const auditContextSchema = z.strictObject({
  ward_id: uuid.nullable(),
  shift_id: uuid.nullable(),
  assignment_ids: z.array(uuid).max(100),
});

export const auditEventSchema = z.strictObject({
  schema_version: z.literal(1),
  event_id: uuid,
  stream_id: uuid,
  sequence: z.number().int().min(1),
  event_type: z.string().min(1).max(80),
  recorded_at: utcDateTime,
  occurred_at: utcDateTime,
  actor_id: uuid.nullable(),
  role_snapshot: roleSchema.nullable(),
  organization_id: uuid.nullable(),
  patient_ref: uuid.nullable(),
  source_org: uuid.nullable(),
  recipient_org: uuid.nullable(),
  resource_domain: domainSchema.nullable(),
  action: z.string().min(1).max(80),
  decision: z.enum(["ALLOW", "DENY", "NOT_APPLICABLE"]),
  reason_code: z.string().min(1).max(80),
  policy_version: z.number().int().min(1).nullable(),
  consent_or_emergency_ref: uuid.nullable(),
  correlation_id: uuid,
  outcome: z.enum(["SUCCEEDED", "DENIED", "ABORTED", "FAILED", "UNKNOWN"]),
  context: auditContextSchema,
  justification_id: uuid.nullable(),
  justification_digest: hash.nullable(),
  previous_hash: hash,
  event_hash: hash,
});

const collectionFields = {
  next_cursor: z.string().min(1).max(2048).nullable(),
  correlation_id: uuid,
  source: sourceSchema.nullable(),
  retrieved_at: utcDateTime,
  completeness_notice: z.string().min(1).max(400),
};

export const eventCollectionSchema = z.strictObject({
  items: z.array(auditEventSchema).max(100),
  ...collectionFields,
});

export const alertRuleSchema = z.enum([
  "AR01",
  "AR02",
  "AR03",
  "AR04",
  "AR05",
  "AR06",
  "AR07",
  "AR08",
  "AR09",
]);

export const securityAlertSchema = z.strictObject({
  id: uuid,
  event_id: uuid,
  stream_id: uuid,
  rule_id: alertRuleSchema,
  severity: z.enum(["HIGH", "CRITICAL"]),
  status: z.enum([
    "REVIEW_REQUIRED",
    "IN_REVIEW",
    "RESOLVED_LEGITIMATE",
    "RESOLVED_SUSPECTED_MISUSE",
  ]),
  actor_id: uuid,
  organization_id: uuid,
  patient_ref: uuid.nullable(),
  reason_code: z.string().min(1).max(80),
  created_at: utcDateTime,
  reviewer_id: uuid.nullable(),
  resolution: z.string().min(20).max(1000).nullable(),
  version: z.number().int().min(1),
});

export const alertCollectionSchema = z.strictObject({
  items: z.array(securityAlertSchema).max(100),
  ...collectionFields,
});

export const reviewAlertSchema = z.strictObject({
  target_status: z.enum(["IN_REVIEW", "RESOLVED_LEGITIMATE", "RESOLVED_SUSPECTED_MISUSE"]),
  explanation: z.string().trim().min(20).max(1000),
  expected_version: z.number().int().min(1),
});

export const alertResponseSchema = z.strictObject({
  alert: securityAlertSchema,
  correlation_id: uuid,
});

export const checkpointSchema = z.strictObject({
  stream_id: uuid,
  sequence: z.number().int().min(0),
  head_hash: hash,
  created_at: utcDateTime,
});

export const verifyChainRequestSchema = z.strictObject({
  trusted_checkpoint_id: uuid.optional(),
});

export const chainVerificationSchema = z.strictObject({
  stream_id: uuid,
  status: z.enum(["VALID", "INVALID"]),
  checked_from: z.number().int().min(0),
  checked_to: z.number().int().min(0),
  first_failing_sequence: z.number().int().min(0).nullable(),
  reason: z
    .enum(["HASH_MISMATCH", "SEQUENCE_GAP", "LINK_MISMATCH", "CHECKPOINT_MISMATCH", "TRUNCATED"])
    .nullable(),
  checkpoint_comparison: z.enum(["MATCH", "MISMATCH", "NOT_PROVIDED"]),
  checkpoint: checkpointSchema,
  verified_at: utcDateTime,
  limitations: z.array(z.string().min(1).max(300)).max(10),
  correlation_id: uuid,
});

export type AuditEvent = z.infer<typeof auditEventSchema>;
export type EventCollection = z.infer<typeof eventCollectionSchema>;
export type AlertRule = z.infer<typeof alertRuleSchema>;
export type SecurityAlert = z.infer<typeof securityAlertSchema>;
export type AlertCollection = z.infer<typeof alertCollectionSchema>;
export type ReviewAlert = z.infer<typeof reviewAlertSchema>;
export type AlertResponse = z.infer<typeof alertResponseSchema>;
export type Checkpoint = z.infer<typeof checkpointSchema>;
export type VerifyChainRequest = z.infer<typeof verifyChainRequestSchema>;
export type ChainVerification = z.infer<typeof chainVerificationSchema>;
export type SecurityRole = Role;
export type SecuritySource = Source;
